array.c

```diff
@@ -468,6 +495,23 @@ rb_ary_detransient(VALUE ary)
 }
 #endif
 
+void
+rb_ary_make_embedded(VALUE ary)
+{
+    assert(rb_ary_pool_moveable_p(ary));
+    if (!ARY_EMBED_P(ary) && !OBJ_FROZEN(ary))
+    {
+        VALUE *buf = RARRAY_PTR(ary);
+        long len = RARRAY_LEN(ary);
+
+        FL_SET_EMBED(ary);
+        ARY_SET_EMBED_LEN(ary, len);;
+
+        memmove(RARRAY_PTR(ary), buf, len * sizeof(VALUE));
+        ary_heap_free_ptr(ary, buf, len * sizeof(VALUE));
```

Comment:
We need to unset the transient flag here since `ary_heap_free_ptr` doesn't do that.